source/core/NumericString.cpp

```diff
@@ -1232,6 +1232,124 @@ Boolean BelongtoCharSet(SubString* charSet, Utf8Char candidate) {
+        case 'x':
+        case 'd':
+        case 'u':
+        case 'b':
+        case 'B':
+        case 'o': {
+                intValue = ScanIntBaseValues (formatChar, beginPointer, endPointer);
+            }
+            break;
+        case 'f':
+        case 'e': {
+                Double temp = strtold(beginPointer, endPointer);
+                // or round to even
+                intValue = (IntMax)temp;
+            }
+            break;
+        default: {
+                intValue = strtoull(beginPointer, endPointer, 10);
+            }
+            break;
+    }
+
+    return intValue;
+}
+
+//----------------------------------------------------------------------------------------------------
+void IntScanString(StaticTypeAndData* argument, TypeRef argumentType, char formatChar, char* beginPointer, char** endPointer)
+{
+    IntMax intValue = ScanIntValues(formatChar, beginPointer, endPointer);
+    intValue = ConvertNumericRange(kEncoding_UInt, argumentType->TopAQSize(), intValue);
+    WriteIntToMemory(argumentType, argument->_pData, intValue);
+}
+
+//----------------------------------------------------------------------------------------------------
+void S2CIntScanString(StaticTypeAndData* argument, TypeRef argumentType, char formatChar, char* beginPointer, char** endPointer)
+{
+    IntMax intValue = ScanIntValues(formatChar, beginPointer, endPointer);
+    intValue = ConvertNumericRange(kEncoding_S2CInt, argumentType->TopAQSize(), intValue);
+    WriteIntToMemory(argumentType, argument->_pData, intValue);
+}
+
+//----------------------------------------------------------------------------------------------------
+void DoubleScanString(StaticTypeAndData* argument, TypeRef argumentType, TempStackCString* truncateInput, char formatChar, char* beginPointer, char** endPointer)
+{
+    double doubleValue;
+    IntMax intValue = 0;
+    switch (formatChar) {
+        case 'x':
+        case 'd':
+        case 'u':
+        case 'b':
+        case 'B':
+        case 'o': {
+                intValue = ScanIntBaseValues(formatChar, beginPointer, endPointer);
+                doubleValue = intValue;
+            }
+            break;
+        case 'f': 
+        case 'e': 
+        case 'g': 
+        case 'p': {
+                doubleValue = strtold(beginPointer, endPointer);
+                if (formatChar == 'p' && *endPointer != NULL && *endPointer < ConstCStr(truncateInput->End())) {
+                    char siPrefixesTable[] = { 'y', 'z', 'a', 'f', 'p', 'n','u', 'm', '0', 'k', 'M','G', 'T', 'P','E', 'Z', 'Y' };
+                    char prefix = **endPointer;
+                    int i = 0;
+                    for (; i < sizeof(siPrefixesTable); ++i)
+                        if (siPrefixesTable[i] == prefix)
+                            break;
+                    if (i < sizeof(siPrefixesTable)) {
+                        ++beginPointer;
+                        doubleValue *= pow(10.0, (i - 8) * 3);
+                    }
```

Comment:
> Moving the declaration of i would invalidate the test of it after the loop to detect if the loop completed without finding the prefix so a default value can be provided

I think this is still handled properly by moving that test into the loop, since the default is to skip scaling, but ...

> Yes, that's too pedantic. sizeof(char) will always be 1.

... yes, you're right -- bits may differ, but sizeof(char) is 1. The more likely scenario is the table changing to another type to hold a different encoding, but again ... unlikely, still too pedantic.